components/service/glean/src/main/java/mozilla/components/service/glean/storages/EventsStorageEngine.kt

```diff
@@ -151,16 +247,85 @@ internal object EventsStorageEngine : StorageEngine {
         return getSnapshot(storeName, clearStore)?.let { pingEvents ->
             val eventsArray = JSONArray()
             pingEvents.forEach {
-                val eventData = JSONArray()
-                eventData.put(it.msSinceStart)
-                eventData.put(it.category)
-                eventData.put(it.name)
-                if (it.extra != null) {
-                    eventData.put(JSONObject(it.extra))
-                }
-                eventsArray.put(eventData)
+                eventsArray.put(serializeEvent(it))
+            }
+            eventsArray
+        }
+    }
+
+    /**
+     * Get the File object in which to store events for the given store.
+     *
+     * @param storeName The name of the store
+     * @return File object to store events
+     */
+    fun getFile(storeName: String): File {
+        return File(storageDirectory, storeName)
+    }
+
+    /**
+     * Serializes an event to its JSON representation.
+     *
+     * @param category Event category
+     * @param name Event name
+     * @param msSinceStart The ms since the process started
+     * @param extra A `Map<String, String>` of extra values
+     * @return JSONArray representing the event
+     */
+    fun serializeEvent(event: RecordedEventData): JSONArray {
+        val eventData = JSONArray()
+        eventData.put(event.timestamp)
+        eventData.put(event.category)
+        eventData.put(event.name)
+        if (event.extra != null) {
+            eventData.put(JSONObject(event.extra))
+        }
+        return eventData
+    }
+
+    /**
+     * Deserializes an event in JSON into a RecordedEventData object.
+     *
+     * @param jsonContent The JSONArray containing the data for the event.
+     * @return event data object
+     */
+    fun deserializeEvent(jsonContent: JSONArray): RecordedEventData {
+        if (jsonContent.length() != N_FIELDS) {
+            throw JSONException("Invalid event")
+        }
+
+        val extra: Map<String, String>? = jsonContent.optJSONObject(EXTRA_FIELD)?.let {
+            var extraValues: MutableMap<String, String> = mutableMapOf()
+            val names = it.names()
+            for (i in 0..(names.length() - 1)) {
+                extraValues[names.getString(i)] = it.getString(names.getString(i))
+            }
+            extraValues
+        }
+
+        return RecordedEventData(
+            jsonContent.getString(CATEGORY_FIELD),
+            jsonContent.getString(NAME_FIELD),
+            jsonContent.getLong(TIMESTAMP_FIELD),
+            extra
+        )
+    }
+
+    /**
+     * Writes an event to a single store on disk.
+     *
+     * @param storeName The name of the store to add the event to.
+     * @param eventContent A string of JSON as returned by [serializeEvent]
+     */
+    @Synchronized
```

Comment:
This is happening on a separate thread and we're not really accessing `eventStore`. Does this need to be `@Synchronized`?